homeassistant/components/tplink/config_flow.py

```diff
@@ -40,27 +66,108 @@ async def async_step_integration_discovery(
+    async def _async_handle_discovery(
+        self, host: str, mac: str, config: Optional[dict] = None
+    ) -> FlowResult:
         """Handle any discovery."""
         await self.async_set_unique_id(dr.format_mac(mac))
+        if config:
+            self._update_config_if_entry_in_setup_retry(config)
         self._abort_if_unique_id_configured(updates={CONF_HOST: host})
         self._async_abort_entries_match({CONF_HOST: host})
         self.context[CONF_HOST] = host
         for progress in self._async_in_progress():
             if progress.get("context", {}).get(CONF_HOST) == host:
                 return self.async_abort(reason="already_in_progress")
-
+        credentials = await get_credentials(self.hass)
         try:
-            self._discovered_device = await self._async_try_connect(
-                host, raise_on_progress=True
+            await self._async_try_discover_and_update(
+                host, credentials, raise_on_progress=True
             )
+        except AuthenticationException:
+            return await self.async_step_discovery_auth_confirm()
         except SmartDeviceException:
             return self.async_abort(reason="cannot_connect")
+
         return await self.async_step_discovery_confirm()
 
+    async def async_step_discovery_auth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Dialog that informs the user that auth is required."""
+        assert self._discovered_device is not None
+        errors = {}
+
+        credentials = await get_credentials(self.hass)
+        if credentials != self._discovered_device.config.credentials:
+            try:
+                device = await self._async_try_connect(
+                    self._discovered_device, credentials
+                )
+            except AuthenticationException:
+                pass  # Authentication exceptions should continue to the rest of the step
+            else:
+                self._discovered_device = device
+                return await self.async_step_discovery_confirm()
+
+        if user_input:
+            credentials = Credentials(
+                user_input[CONF_USERNAME], user_input[CONF_PASSWORD]
+            )
+            try:
+                device = await self._async_try_connect(
+                    self._discovered_device, credentials
+                )
+            except AuthenticationException:
+                errors["base"] = "invalid_auth"
+            except SmartDeviceException:
+                return self.async_abort(reason="cannot_connect")
+            else:
+                self._discovered_device = device
+                await set_credentials(
+                    self.hass, user_input[CONF_USERNAME], user_input[CONF_PASSWORD]
+                )
+                async_call_later(
+                    self.hass,
+                    REAUTH_REFLOW_CALL_LATER_SECS,
+                    self._async_reload_requires_auth_entries,
+                )
```

Comment:
This looks brittle.  Why use a call later?

Reply:
Done using `hass.async_create_task`